src/osd/OSD.cc

```diff
@@ -8308,63 +8311,72 @@ void OSD::handle_osd_map(MOSDMap *m)
 }
 
 /*
- *  For each added map which was sent in this MOSDMap [first, last]:
- *
- *  1) Check if a pool was deleted
- *  2) For existing pools, check if pg_num was changed
- *  3) Check if a pool was created
- *
- *  Track all of the changes above in pg_num_history
+ *  Compare between the previous last_map to the current one
+ *  Track all of the changes relevant in pg_num_history
  */
 void OSD::track_pools_and_pg_num_changes(map<epoch_t,OSDMapRef>& added_maps,
                                          ObjectStore::Transaction& t,
+                                         epoch_t first,
                                          epoch_t last)
 {
-  OSDMapRef lastmap;
-  for (auto& [added_map_epoch, added_map] : added_maps) {
-    if (!lastmap) {
-      if (!(lastmap = service.try_get_map(added_map_epoch - 1))) {
-        dout(10) << __func__ << " can't get previous map " << added_map_epoch - 1
-                 << " probably first start of this osd" << dendl;
-        continue;
-      }
+  OSDMapRef current_lastmap = added_maps[last];
+  OSDMapRef previous_lastmap;
+
+  if (superblock.maps.empty()) {
+    dout(10) << __func__ << " no maps stored, this is probably "
+             << "the first start of this osd" << dendl;
+    previous_lastmap = added_maps[first];
+  } else {
+    if (first > superblock.get_newest_map() + 1) {
+      ceph_assert(first == superblock.cluster_osdmap_trim_lower_bound);
+      dout(10) << __func__ << " can't get previous map " << first - 1
+               << " first start of this osd after a map gap" << dendl;
```

Comment:
The map we actually want isn't first - 1, it's superblock.get_newest_map().  This should probably be dout(20) and should simply note that the osd has restarted after a map gap.  It doesn't impact behavior, just a useful thing to have in the log.